### PR TITLE
Add State Views and Config View specialization 

### PR DESCRIPTION
**Pull request by peterschwarz**
**Add StateView and its Factory**

StateViewFactory provides read-only views into state stored in a merkle tree database. These views can provide a window into the state which can be used throughout a block of processing without concern of the state being changed underneath the process.

**Add ConfigView and its Factory**

ConfigView is a specialized StateView which translates keys into state addresses and transforms state entries into their respective values. This can be used by validator components to read on-chain settings at viewed from the lens of a specific point in the chain.

### CODE REVIEW

**Comment by peterschwarz** [comment]
This is a reopen of #85

File: validator/tests/unit3/test_state_view/__init__.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# Copyright 2016 Intel Corporation
```
Comment (vaporos) [inline validator/tests/unit3/test_state_view/__init__.py]: 2017

File: validator/tests/unit3/test_state_view/tests.py [hunk 2]
```diff
@@ -0,0 +1,53 @@
+# copyright 2017 intel corporation
```
Comment (vaporos) [inline validator/tests/unit3/test_state_view/tests.py]: Capitalize copyright

Comment (vaporos) [inline validator/tests/unit3/test_state_view/tests.py]: This entire block is lower cased

File: validator/tests/unit3/test_state_view/tests.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+    def test_state_view(self):
+        merkle_db = MerkleDatabase(self._database)
+
+        state_view_factory = StateViewFactory(self._database)
+
+        initial_state_view = state_view_factory.create_view(
+            merkle_db.get_merkle_root())
+
+        # test that the intial state view returns empty values
```
Comment (vaporos) [inline validator/tests/unit3/test_state_view/tests.py]: Typo intial

File: validator/sawtooth_validator/state/state_view.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        """Creates a StateView for the given state root hash.
+
+        Returns:
+            StateView: state view locked to the given root hash.
+        """
+        return StateView(MerkleDatabase(self._database, state_root_hash))
+
+
+class StateView(object):
+    """The StateView is a snapshot of state stored in a merkle tree.
```
Comment (cintel) [inline validator/sawtooth_validator/state/state_view.py]: Perhaps: "The StateView provides read only access to the state stored at a particular merkle root."

File: validator/sawtooth_validator/state/config_view.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+# ------------------------------------------------------------------------------
+
+import hashlib
+
+from sawtooth_validator.protobuf.setting_pb2 import Setting
+
+
+class ConfigView(object):
+    """
+    A ConfigView is a snapshot view of on-chain configuration settings.
```
Comment (cintel) [inline validator/sawtooth_validator/state/config_view.py]: Perhaps: "A ConfigView provides access to on-chain configuration stored at a particular merkle root.".

File: validator/sawtooth_validator/state/config_view.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    """
+
+    def __init__(self, state_view):
+        """Creates a ConfigView, given a StateView as its snapshot.
```
Comment (cintel) [inline validator/sawtooth_validator/state/config_view.py]: Creates a ConfigView, given a StateView for merkle tree access.

**Pull request merged**